runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -643,4 +653,24 @@ api.nvim_set_decoration_provider(namespace, {
 ---@private
 M.__STHighlighter = STHighlighter
 
+--- Return the token under the cursor
+---@private
+---@param winnr (number|nil) Window handle or 0 for current window (default)
+---
+---@return string[]|nil token
+M.get_token_at_cursor = function(winnr)
+  winnr = winnr or 0
+  local cursor = api.nvim_win_get_cursor(winnr)
+
+  local highlights = STHighlighter.active[1].client_state[1].current_result.highlights
```

Comment:
You can reuse the binary search method I have in here so you don't have to iterate every highlight. `on_win` has the example usage.